File: results/0.25/ELF/elf-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

#define EI_NIDENT 16

typedef enum {
    ET_NONE = 0,
    ET_REL = 1,
    ET_EXEC = 2,
    ET_DYN = 3,
    ET_CORE = 4
} Elf_Type;

typedef enum {
    EM_NONE = 0,
    EM_M32 = 1,
    EM_SPARC = 2,
    EM_386 = 3,
    EM_68K = 4,
    EM_88K = 5,
    EM_860 = 7,
    EM_MIPS = 8,
    EM_PARISC = 15,
    EM_PPC = 20,
    EM_PPC64 = 21,
    EM_S390 = 22,
    EM_SPU = 23,
    EM_V800 = 36,
    EM_FR20 = 37,
    EM_RH32 = 38,
    EM_RCE = 39,
    EM_ARM = 40,
    EM_ALPHA = 41,
    EM_SH = 42,
    EM_SPARCV9 = 43,
    EM_TRICORE = 44,
    EM_ARC = 45,
    EM_H8_300 = 46,
    EM_H8_300H = 47,
    EM_H8S = 48,
    EM_H8_500 = 49,
    EM_IA_64 = 50,
    EM_MIPS_X = 51,
    EM_COLDFIRE = 52,
    EM_68HC12 = 53,
    EM_MMA = 54,
    EM_PCP = 55,
    EM_NCPU = 56,
    EM_NDR1 = 57,
    EM_STARCORE = 58,
    EM_ME16 = 59,
    EM_ST100 = 60,
    EM_TINYJ = 61,
    EM_X86_64 = 62,
    EM_PDSP = 63,
    EM_PDP10 = 64,
    EM_PDP11 = 65,
    EM_FX66 = 66,
    EM_ST9PLUS = 67,
    EM_ST7 = 68,
    EM_68HC16 = 69,
    EM_68HC11 = 70,
    EM_68HC08 = 71,
    EM_68HC05 = 72,
    EM_SVX = 73,
    EM_ST19 = 74,
    EM_VAX = 75,
    EM_CRIS = 76,
    EM_JAVELIN = 77,
    EM_FIREPATH = 78,
    EM_ZSP = 79,
    EM_MMIX = 80,
    EM_HUANY = 81,
    EM_PRISM = 82,
    EM_AVR = 83,
    EM_FR30 = 84,
    EM_D10V = 85,
    EM_D30V = 86,
    EM_V850 = 87,
    EM_M32R = 88,
    EM_MN10300 = 89,
    EM_MN10200 = 90,
    EM_PJ = 91,
    EM_OPENRISC = 92,
    EM_ARC_A5 = 93,
    EM_XTENSA = 94,
    EM_NUM = 95
} Elf_Machine;

typedef enum {
    PT_NULL = 0,
    PT_LOAD = 1,
    PT_DYNAMIC = 2,
    PT_INTERP = 3,
    PT_NOTE = 4,
    PT_SHLIB = 5,
    PT_PHDR = 6,
    PT_TLS = 7,
    PT_NUM = 8,
    PT_LOOS = 0x60000000,
    PT_HIOS = 0x6fffffff,
    PT_LOPROC = 0x70000000,
    PT_HIPROC = 0x7fffffff
} Elf_Phdr_Type;

typedef enum {
    SHT_NULL = 0,
    SHT_PROGBITS = 1,
    SHT_SYMTAB = 2,
    SHT_STRTAB = 3,
    SHT_RELA = 4,
    SHT_HASH = 5,
    SHT_DYNAMIC = 6,
    SHT_NOTE = 7,
    SHT_NOBITS = 8,
    SHT_REL = 9,
    SHT_SHLIB = 10,
    SHT_DYNSYM = 11,
    SHT_INIT_ARRAY = 14,
    SHT_FINI_ARRAY = 15,
    SHT_PREINIT_ARRAY = 16,
    SHT_GROUP = 17,
    SHT_SYMTAB_SHNDX = 18,
    SHT_NUM = 19,
    SHT_LOOS = 0x60000000,
    SHT_HIOS = 0x6fffffff,
    SHT_LOPROC = 0x70000000,
    SHT_HIPROC = 0x7fffffff,
    SHT_LOUSER = 0x80000000,
    SHT_HIUSER = 0x8fffffff
} Elf_Shdr_Type;

typedef enum {
    SHF_WRITE = 0x1,
    SHF_ALLOC = 0x2,
    SHF_EXECINSTR = 0x4,
    SHF_MERGE = 0x10,
    SHF_STRINGS = 0x20,
    SHF_INFO_LINK = 0x40,
    SHF_LINK_ORDER = 0x80,
    SHF_OS_NONCONFORMING = 0x100,
    SHF_GROUP = 0x200,
    SHF_TLS = 0x400,
    SHF_MASKOS = 0x0ff00000,
    SHF_MASKPROC = 0xf0000000
} Elf_Shdr_Flag;

typedef enum {
    STB_LOCAL = 0,
    STB_GLOBAL = 1,
    STB_WEAK = 2,
    STB_NUM = 3,
    STB_LOOS = 10,
    STB_HIOS = 12,
    STB_LOPROC = 13,
    STB_HIPROC = 15
} Elf_Sym_Binding;

typedef enum {
    STT_NOTYPE = 0,
    STT_OBJECT = 1,
    STT_FUNC = 2,
    STT_SECTION = 3,
    STT_FILE = 4,
    STT_COMMON = 5,
    STT_TLS = 6,
    STT_NUM = 7,
    STT_LOOS = 10,
    STT_HIOS = 12,
    STT_LOPROC = 13,
    STT_HIPROC = 15
} Elf_Sym_Type;

typedef enum {
    DT_NULL = 0,
    DT_NEEDED = 1,
    DT_PLTRELSZ = 2,
    DT_PLTGOT = 3,
    DT_HASH = 4,
    DT_STRTAB = 5,
    DT_SYMTAB = 6,
    DT_RELA = 7,
    DT_RELASZ = 8,
    DT_RELAENT = 9,
    DT_STRSZ = 10,
    DT_SYMENT = 11,
    DT_INIT = 12,
    DT_FINI = 13,
    DT_SONAME = 14,
    DT_RPATH = 15,
    DT_SYMBOLIC = 16,
    DT_REL = 17,
    DT_RELSZ = 18,
    DT_RELENT = 19,
    DT_PLTREL = 20,
    DT_DEBUG = 21,
    DT_TEXTREL = 22,
    DT_JMPREL = 23,
    DT_BIND_NOW = 24,
    DT_INIT_ARRAY = 25,
    DT_FINI_ARRAY = 26,
    DT_INIT_ARRAYSZ = 27,
    DT_FINI_ARRAYSZ = 28,
    DT_RUNPATH = 29,
    DT_FLAGS = 30,
    DT_PREINIT_ARRAY = 32,
    DT_PREINIT_ARRAYSZ = 33,
    DT_NUM = 34,
    DT_LOOS = 0x60000000,
    DT_HIOS = 0x6fffffff,
    DT_LOPROC = 0x70000000,
    DT_HIPROC = 0x7fffffff
} Elf_Dyn_Tag;

typedef struct {
    unsigned char e_ident[EI_NIDENT];
    Elf_Type e_type;
    Elf_Machine e_machine;
    unsigned int e_version;
    unsigned long e_entry;
    unsigned long e_phoff;
    unsigned long e_shoff;
    unsigned int e_flags;
    unsigned short e_ehsize;
    unsigned short e_phentsize;
    unsigned short e_phnum;
    unsigned short e_shentsize;
    unsigned short e_shnum;
    unsigned short e_shstrndx;
} Elf32_Ehdr;

typedef struct {
    unsigned char e_ident[EI_NIDENT];
    Elf_Type e_type;
    Elf_Machine e_machine;
    unsigned int e_version;
    unsigned long long e_entry;
    unsigned long long e_phoff;
    unsigned long long e_shoff;
    unsigned int e_flags;
    unsigned short e_ehsize;
    unsigned short e_phentsize;
    unsigned short e_phnum;
    unsigned short e_shentsize;
    unsigned short e_shnum;
    unsigned short e_shstrndx;
} Elf64_Ehdr;

typedef struct {
    Elf_Phdr_Type p_type;
    unsigned long p_offset;
    unsigned long p_vaddr;
    unsigned long p_paddr;
    unsigned long p_filesz;
    unsigned long p_memsz;
    unsigned long p_flags;
    unsigned long p_align;
} Elf32_Phdr;

typedef struct {
    Elf_Phdr_Type p_type;
    unsigned long long p_offset;
    unsigned long long p_vaddr;
    unsigned long long p_paddr;
    unsigned long long p_filesz;
    unsigned long long p_memsz;
    unsigned long p_flags;
    unsigned long long p_align;
} Elf64_Phdr;

typedef struct {
    unsigned int sh_name;
    Elf_Shdr_Type sh_type;
    Elf_Shdr_Flag sh_flags;
    unsigned long sh_addr;
    unsigned long sh_offset;
    unsigned long sh_size;
    unsigned int sh_link;
    unsigned int sh_info;
    unsigned long sh_addralign;
    unsigned long sh_entsize;
} Elf32_Shdr;

typedef struct {
    unsigned int sh_name;
    Elf_Shdr_Type sh_type;
    Elf_Shdr_Flag sh_flags;
    unsigned long long sh_addr;
    unsigned long long sh_offset;
    unsigned long long sh_size;
    unsigned int sh_link;
    unsigned int sh_info;
    unsigned long long sh_addralign;
    unsigned long long sh_entsize;
} Elf64_Shdr;

typedef struct {
    unsigned int st_name;
    unsigned char st_info;
    unsigned char st_other;
    unsigned short st_shndx;
    unsigned long st_value;
    unsigned long st_size;
} Elf32_Sym;

typedef struct {
    unsigned int st_name;
    unsigned char st_info;
    unsigned char st_other;
    unsigned short st_shndx;
    unsigned long long st_value;
    unsigned long long st_size;
} Elf64_Sym;

typedef struct {
    unsigned int d_tag;
    unsigned long d_un;
} Elf32_Dyn;

typedef struct {
    unsigned int d_tag;
    unsigned long long d_un;
} Elf64_Dyn;

typedef struct {
    unsigned int r_offset;
    unsigned int r_info;
} Elf32_Rel;

typedef struct {
    unsigned long long r_offset;
    unsigned int r_info;
    unsigned long long r_addend;
} Elf64_Rel;

typedef struct {
    unsigned long long r_offset;
    unsigned int r_info;
    unsigned long long r_addend;
} Elf64_Rela;

HParser* HParser_new(void) {
    HParser* parser = malloc(sizeof(HParser));
    return parser;
}

void HParser_set_input(HParser* parser, FILE* file) {
    parser->file = file;
}

void HParser_def(HParser* parser, const char* name, HParser* definition) {
    parser->definitions[name] = definition;
}

HParser* HParser_seq(HParser* parser, ...) {
    va_list args;
    va_start(args, parser);
    HParser* seq = malloc(sizeof(HParser));
    seq->type = HParser_SEQ;
    seq->args = args;
    va_end(args);
    return seq;
}

HParser* HParser_bytes(int size) {
    HParser* bytes = malloc(sizeof(HParser));
    bytes->type = HParser_BYTES;
    bytes->size = size;
    return bytes;
}

HParser* HParser_uint16(void) {
    HParser* uint16 = malloc(sizeof(HParser));
    uint16->type = HParser_UINT16;
    return uint16;
}

HParser* HParser_uint32(void) {
    HParser* uint32 = malloc(sizeof(HParser));
    uint32->type = HParser_UINT32;
    return uint32;
}

HParser* HParser_uint64(void) {
    HParser* uint64 = malloc(sizeof(HParser));
    uint64->type = HParser_UINT64;
    return uint64;
}

HParser* HParser_uint8(void) {
    HParser* uint8 = malloc(sizeof(HParser));
    uint8->type = HParser_UINT8;
    return uint8;
}

HParser* HParser_call(const char* name) {
    HParser* call = malloc(sizeof(HParser));
    call->type = HParser_CALL;
    call->name = name;
    return call;
}

HParser* HParser_repeat(HParser* parser, HParser* count) {
    HParser* repeat = malloc(sizeof(HParser));
    repeat->type = HParser_REPEAT;
    repeat->parser = parser;
    repeat->count = count;
    return repeat;
}

void HParser_parse(HParser* parser, const char* name) {
    // implementation of HParser_parse
}

void HParser_free(HParser* parser) {
    free(parser);
}

int main(int argc, char **argv) {
    if (argc != 2) {
        printf("Usage: %s <elf_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        printf("Error opening file %s\n", argv[1]);
        return 1;
    }

    HParser *parser = HParser_new();
    HParser_set_input(parser, file);

    HParser_def(parser, "elf32_ehdr", HParser_seq(
        HParser_bytes(EI_NIDENT),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16()
    ));

    HParser_def(parser, "elf64_ehdr", HParser_seq(
        HParser_bytes(EI_NIDENT),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint32(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint32(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16(),
        HParser_uint16()
    ));

    HParser_def(parser, "elf32_phdr", HParser_seq(
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32()
    ));

    HParser_def(parser, "elf64_phdr", HParser_seq(
        HParser_uint32(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint32(),
        HParser_uint64()
    ));

    HParser_def(parser, "elf32_shdr", HParser_seq(
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint32()
    ));

    HParser_def(parser, "elf64_shdr", HParser_seq(
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint64(),
        HParser_uint32(),
        HParser_uint32(),
        HParser_uint64(),
        HParser_uint64()
    ));

    HParser_def(parser, "elf32_sym", HParser_seq(
        HParser_uint32(),
        HParser_uint8(),
        HParser_uint8(),
        HParser_uint16(),
        HParser_uint32(),
        HParser_uint32()
    ));

    HParser_def(parser, "elf64_sym", HParser_seq(
        HParser_uint32(),
        HParser_uint8(),
        HParser_uint8(),
        HParser_uint16(),
        HParser_uint64(),
        HParser_uint64()
    ));

    HParser_def(parser, "elf32_dyn", HParser_seq(
        HParser_uint32(),
        HParser_uint32()
    ));

    HParser_def(parser, "elf64_dyn", HParser_seq(
        HParser_uint32(),
        HParser_uint64()
    ));

    HParser_def(parser, "elf32_rel", HParser_seq(
        HParser_uint32(),
        HParser_uint32()
    ));

    HParser_def(parser, "elf64_rel", HParser_seq(
        HParser_uint64(),
        HParser_uint32(),
        HParser_uint64()
    ));

    HParser_def(parser, "elf64_rela", HParser_seq(
        HParser_uint64(),
        HParser_uint32(),
        HParser_uint64()
    ));

    HParser_def(parser, "elf32", HParser_seq(
        HParser_call("elf32_ehdr"),
        HParser_repeat(HParser_call("elf32_phdr"), HParser_uint16()),
        HParser_repeat(HParser_call("elf32_shdr"), HParser_uint16())
    ));

    HParser_def(parser, "elf64", HParser_seq(
        HParser_call("elf64_ehdr"),
        HParser_repeat(HParser_call("elf64_phdr"), HParser_uint16()),
        HParser_repeat(HParser_call("elf64_shdr"), HParser_uint16())
    ));

    HParser_parse(parser, "elf32");
    HParser_parse(parser, "elf64");

    fclose(file);
    HParser_free(parser);
    return 0;
}

File: results/0.25/ELF/elf-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    bytes::complete::{take},
    combinator::{map},
    multi::{count},
    number::complete::{be_u16, be_u32, be_u64, be_u8},
    IResult,
};
use std::{
    env,
    fs::File,
    io::{BufReader, Read},
};

#[derive(Debug)]
enum ElfClass {
    Elf32,
    Elf64,
}

#[derive(Debug)]
enum ElfData {
    LittleEndian,
    BigEndian,
}

#[derive(Debug)]
enum OsAbi {
    SystemV,
    HpUx,
    NetBSD,
    Linux,
    Other(u8),
}

#[derive(Debug)]
enum ElfType {
    Rel,
    Exec,
    Dyn,
    Core,
}

#[derive(Debug)]
enum Machine {
    M32,
    Sparc,
    I386,
    Other(u16),
}

#[derive(Debug)]
struct ElfHeader {
    e_ident: [u8; 16],
    e_type: ElfType,
    e_machine: Machine,
    e_version: u32,
    e_entry: u64,
    e_phoff: u64,
    e_shoff: u64,
    e_flags: u32,
    e_ehsize: u16,
    e_phentsize: u16,
    e_phnum: u16,
    e_shentsize: u16,
    e_shnum: u16,
    e_shstrndx: u16,
}

fn parse_elf_class(input: &[u8]) -> IResult<&[u8], ElfClass> {
    map(be_u8, |x| match x {
        1 => ElfClass::Elf32,
        2 => ElfClass::Elf64,
        _ => panic!("Invalid ELF class"),
    })(input)
}

fn parse_elf_data(input: &[u8]) -> IResult<&[u8], ElfData> {
    map(be_u8, |x| match x {
        1 => ElfData::LittleEndian,
        2 => ElfData::BigEndian,
        _ => panic!("Invalid ELF data encoding"),
    })(input)
}

fn parse_os_abi(input: &[u8]) -> IResult<&[u8], OsAbi> {
    map(be_u8, |x| match x {
        0 => OsAbi::SystemV,
        1 => OsAbi::HpUx,
        2 => OsAbi::NetBSD,
        3 => OsAbi::Linux,
        x => OsAbi::Other(x),
    })(input)
}

fn parse_elf_type(input: &[u8]) -> IResult<&[u8], ElfType> {
    map(be_u16, |x| match x {
        1 => ElfType::Rel,
        2 => ElfType::Exec,
        3 => ElfType::Dyn,
        4 => ElfType::Core,
        _ => panic!("Invalid ELF type"),
    })(input)
}

fn parse_machine(input: &[u8]) -> IResult<&[u8], Machine> {
    map(be_u16, |x| match x {
        1 => Machine::M32,
        2 => Machine::Sparc,
        3 => Machine::I386,
        x => Machine::Other(x),
    })(input)
}

fn parse_elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    let (input, e_ident) = take(16usize)(input)?;
    let e_ident = {
        let mut arr = [0u8; 16];
        arr.copy_from_slice(e_ident);
        arr
    };
    let (input, e_type) = parse_elf_type(input)?;
    let (input, e_machine) = parse_machine(input)?;
    let (input, e_version) = be_u32(input)?;
    let (input, e_entry) = be_u64(input)?;
    let (input, e_phoff) = be_u64(input)?;
    let (input, e_shoff) = be_u64(input)?;
    let (input, e_flags) = be_u32(input)?;
    let (input, e_ehsize) = be_u16(input)?;
    let (input, e_phentsize) = be_u16(input)?;
    let (input, e_phnum) = be_u16(input)?;
    let (input, e_shentsize) = be_u16(input)?;
    let (input, e_shnum) = be_u16(input)?;
    let (input, e_shstrndx) = be_u16(input)?;
    Ok((
        input,
        ElfHeader {
            e_ident,
            e_type,
            e_machine,
            e_version,
            e_entry,
            e_phoff,
            e_shoff,
            e_flags,
            e_ehsize,
            e_phentsize,
            e_phnum,
            e_shentsize,
            e_shnum,
            e_shstrndx,
        },
    ))
}

#[derive(Debug)]
enum ProgramHeaderType {
    Null,
    Load,
    Dynamic,
    Other(u32),
}

#[derive(Debug)]
struct ProgramHeader {
    p_type: ProgramHeaderType,
    p_offset: u64,
    p_vaddr: u64,
    p_paddr: u64,
    p_filesz: u64,
    p_memsz: u64,
    p_flags: u32,
    p_align: u64,
}

fn parse_program_header_type(input: &[u8]) -> IResult<&[u8], ProgramHeaderType> {
    map(be_u32, |x| match x {
        0 => ProgramHeaderType::Null,
        1 => ProgramHeaderType::Load,
        2 => ProgramHeaderType::Dynamic,
        x => ProgramHeaderType::Other(x),
    })(input)
}

fn parse_program_header(input: &[u8]) -> IResult<&[u8], ProgramHeader> {
    let (input, p_type) = parse_program_header_type(input)?;
    let (input, p_offset) = be_u64(input)?;
    let (input, p_vaddr) = be_u64(input)?;
    let (input, p_paddr) = be_u64(input)?;
    let (input, p_filesz) = be_u64(input)?;
    let (input, p_memsz) = be_u64(input)?;
    let (input, p_flags) = be_u32(input)?;
    let (input, p_align) = be_u64(input)?;
    Ok((
        input,
        ProgramHeader {
            p_type,
            p_offset,
            p_vaddr,
            p_paddr,
            p_filesz,
            p_memsz,
            p_flags,
            p_align,
        },
    ))
}

#[derive(Debug)]
enum SectionHeaderType {
    Null,
    Progbits,
    Symtab,
    Other(u32),
}

#[derive(Debug)]
struct SectionHeader {
    sh_name: u32,
    sh_type: SectionHeaderType,
    sh_flags: u64,
    sh_addr: u64,
    sh_offset: u64,
    sh_size: u64,
    sh_link: u32,
    sh_info: u32,
    sh_addralign: u64,
    sh_entsize: u64,
}

fn parse_section_header_type(input: &[u8]) -> IResult<&[u8], SectionHeaderType> {
    map(be_u32, |x| match x {
        0 => SectionHeaderType::Null,
        1 => SectionHeaderType::Progbits,
        2 => SectionHeaderType::Symtab,
        x => SectionHeaderType::Other(x),
    })(input)
}

fn parse_section_header(input: &[u8]) -> IResult<&[u8], SectionHeader> {
    let (input, sh_name) = be_u32(input)?;
    let (input, sh_type) = parse_section_header_type(input)?;
    let (input, sh_flags) = be_u64(input)?;
    let (input, sh_addr) = be_u64(input)?;
    let (input, sh_offset) = be_u64(input)?;
    let (input, sh_size) = be_u64(input)?;
    let (input, sh_link) = be_u32(input)?;
    let (input, sh_info) = be_u32(input)?;
    let (input, sh_addralign) = be_u64(input)?;
    let (input, sh_entsize) = be_u64(input)?;
    Ok((
        input,
        SectionHeader {
            sh_name,
            sh_type,
            sh_flags,
            sh_addr,
            sh_offset,
            sh_size,
            sh_link,
            sh_info,
            sh_addralign,
            sh_entsize,
        },
    ))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <elf_file>", args[0]);
        return;
    }
    let file = File::open(&args[1]).unwrap();
    let mut reader = BufReader::new(file);
    let mut buffer = Vec::new();
    reader.read_to_end(&mut buffer).unwrap();
    let (_input, elf_header) = parse_elf_header(&buffer).unwrap();
    println!("ELF Header: {:?}", elf_header);
    let (_input, program_headers) = count(parse_program_header, elf_header.e_phnum as usize)(&buffer[elf_header.e_phoff as usize..]).unwrap();
    println!("Program Headers: {:?}", program_headers);
    let (_input, section_headers) = count(parse_section_header, elf_header.e_shnum as usize)(&buffer[elf_header.e_shoff as usize..]).unwrap();
    println!("Section Headers: {:?}", section_headers);
}